### PR TITLE
Add clipped Rectifier Linear Unit

### PR DESCRIPTION
**Pull request by jnishi**
I impremented clipped Rectifier Linear Unit, which is written as below.
```ClippedReLU(x, z) = min{max{0,x},z}.```
 `z` is a parameter to cap return value of ReLU.

### CODE REVIEW

**Comment by unnonouno** [comment]
Please use flake8 and fix errors before send PR.
http://docs.chainer.org/en/latest/contribution.html

**Comment by jnishi** [comment]
Fixes erros of flake8.

File: chainer/functions/clipped_relu.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+            'relu_bwd')(gx, x[0], gy[0], self.cap)
+        return gx,
+
+
+def clipped_relu(x, z=20):
+    """Clipped Rectifier Unit function :math:`CReLU(x, z) = min{max{0,x},z}`
+
+    Args:
+        x (~chainer.Variable): Input variable.
+        z (integer): clipping value. (default = 20)
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: s/integer/int/

Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: Is `z` really int?

File: chainer/functions/clipped_relu.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        gx = cuda.empty_like(x[0])
+        cuda.elementwise(
+            'float* gx, const float* x, const float* gy, const float z',
+            'gx[i] = ((x[i] > 0) and (x[i] < z))? gy[i] : 0',
+            'relu_bwd')(gx, x[0], gy[0], self.cap)
+        return gx,
+
+
+def clipped_relu(x, z=20):
+    """Clipped Rectifier Unit function :math:`CReLU(x, z) = min{max{0,x},z}`
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: Use `\min` and `\max`.
Replace `{` with `(`.
Insert a space after `,`.

File: chainer/functions/clipped_relu.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    """clipped ReLU function. clipped ReLU is written as below,
+
+    :math:`ClippedReLU(x, z) = min{max{0,x},z}`
+    z is a parameter to cap return value of ReLU.
+
+    """
+
+    def __init__(self, z):
+        self.cap = z
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: You need to check type of `z`.

File: tests/functions_tests/test_clipped_relu.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+if cuda.available:
+    cuda.init()
+
+
+class TestCReLU(unittest.TestCase):
+
+    def setUp(self):
+        # Avoid unstability of numerical grad
+        self.x = numpy.random.uniform(.5, 1, (3, 2)).astype(numpy.float32)
```
Comment (unnonouno) [inline tests/functions_tests/test_clipped_relu.py]: Is this really required?
Is `uniform(-1, 1, (3, 2))` enough?

File: tests/functions_tests/test_clipped_relu.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+from chainer import functions
+from chainer import gradient_check
+from chainer import testing
+from chainer.testing import attr
+
+if cuda.available:
+    cuda.init()
+
+
+class TestCReLU(unittest.TestCase):
```
Comment (unnonouno) [inline tests/functions_tests/test_clipped_relu.py]: Do you have a forward test?

Comment (unnonouno) [inline tests/functions_tests/test_clipped_relu.py]: Rename `TestCReLU` to `TestClippedReLU`.

**Comment by unnonouno** [comment]
What happens when `x` is 0-dimension array e.g. `x = np.array(0, np.float32)`? Check `dtype` of a return value of forward computation.

**Comment by jnishi** [comment]
Thanks for a lot of comments. I fixed comments you pointed. Please check them again.

> What happens when x is 0-dimension array e.g. x = np.array(0, np.float32)? Check dtype of a return value of forward computation.

Add type check of `x` to refuse 0-dimensional array.

File: chainer/functions/clipped_relu.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
+from chainer import cuda
+from chainer import function
+from chainer.utils import type_check
+import numpy
+
+
+class ClippedReLU(function.Function):
+
+    """clipped ReLU function. clipped ReLU is written as below,
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: Please write a brief explanation of the class in the first line, and write the details after that.

File: chainer/functions/clipped_relu.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+    def forward_gpu(self, x):
+        return cuda.gpuarray.minimum(cuda.gpuarray.maximum(0, x[0]), self.cap),
+
+    def backward_gpu(self, x, gy):
+        gx = cuda.empty_like(x[0])
+        cuda.elementwise(
+            'float* gx, const float* x, const float* gy, const float z',
+            'gx[i] = ((x[i] > 0) and (x[i] < z))? gy[i] : 0',
+            'relu_bwd')(gx, x[0], gy[0], self.cap)
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: Use a unique function name such as `'clipped_relu_bwd'`.

File: tests/functions_tests/test_clipped_relu.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+class TestClippedReLU(unittest.TestCase):
+
+    def setUp(self):
+        self.x = numpy.random.uniform(-1, 1, (3, 2)).astype(numpy.float32)
+        self.gy = numpy.random.uniform(-1, 1, (3, 2)).astype(numpy.float32)
+        self.z = 0.75
+
+    def check_forward(self, x_data):
+        x = chainer.Variable(x_data)
+        y = functions.clipped_relu(x, self.z)
```
Comment (unnonouno) [inline tests/functions_tests/test_clipped_relu.py]: Check the type of `y`. I mean,
```
self.assertEqual(y.data.dtype, numpy.float32)
```

Sometimes a result of a function becomes other types such as `numpy.float64` by mistake.

File: tests/functions_tests/test_clipped_relu.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+from chainer.testing import attr
+
+if cuda.available:
+    cuda.init()
+
+
+class TestClippedReLU(unittest.TestCase):
+
+    def setUp(self):
+        self.x = numpy.random.uniform(-1, 1, (3, 2)).astype(numpy.float32)
```
Comment (unnonouno) [inline tests/functions_tests/test_clipped_relu.py]: Please make a test for 0-dimension array like this:
```
class TestClippedReLUZeroDim(TestClippedReLU):
    def setUp(self):
        self.x = numpy.random.uniform(-1, 1, ()).astype(numpy.float32)
        self.gy = numpy.random.uniform(-1, 1, ()).astype(numpy.float32)
        self.z = 0.75
```

File: chainer/functions/clipped_relu.py (hunk 6 above)
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: Capitalize the first word of the sentences.

File: chainer/functions/clipped_relu.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+from chainer import function
+from chainer.utils import type_check
+import numpy
+
+
+class ClippedReLU(function.Function):
+
+    """clipped ReLU function. clipped ReLU is written as below,
+
+    :math:`ClippedReLU(x, z) = \min(\max(0, x) ,z)`
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: ` ,` --> `, `

File: chainer/functions/clipped_relu.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    z is a parameter to cap return value of ReLU.
+
+    """
+
+    def __init__(self, z):
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: Do you need to check the value of `z`? Is it need to be positive, or all values are acceptable?

File: chainer/functions/clipped_relu.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+        gx = cuda.empty_like(x[0])
+        cuda.elementwise(
+            'float* gx, const float* x, const float* gy, const float z',
+            'gx[i] = ((x[i] > 0) and (x[i] < z))? gy[i] : 0',
+            'relu_bwd')(gx, x[0], gy[0], self.cap)
+        return gx,
+
+
+def clipped_relu(x, z=20.0):
+    """Clipped Rectifier Unit function :math:`CReLU(x, z) = \min(\max(0, x), z)`
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: Please split the brief explanation and the detail. I mean for example:
```
"""Clipped Rectifier Unit function.

 This function is expressed as :math:`f(x, z) = \min(\max(0, x), z)`, where :math:`z` is a clipping value.
...
```

File: chainer/functions/clipped_relu.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+        if not isinstance(z, float):
+            raise TypeError('z must be float value')
+        self.cap = z
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 1)
+        x_type, = in_types
+        type_check.expect(
+            x_type.dtype == numpy.float32,
+            x_type.ndim > 0
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: Is this function really for n(>0)-dimension? I think 0-dim can be acceptable.

File: chainer/functions/clipped_relu.py [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+from chainer.utils import type_check
+import numpy
+
+
+class ClippedReLU(function.Function):
+
+    """clipped ReLU function. clipped ReLU is written as below,
+
+    :math:`ClippedReLU(x, z) = \min(\max(0, x) ,z)`
+    z is a parameter to cap return value of ReLU.
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: use `:math:` for z

**Comment by jnishi** [comment]
Thanks for a lot of comments. I fixed description of clipped relu, support 0-dimentional array, and type check. Please check again.

File: chainer/functions/__init__.py [hunk 15]
```diff
@@ -40,6 +41,7 @@
     nonparameterized_linear.NonparameterizedLinear
 Concat = concat.Concat
 Copy = copy.Copy
+ClippedReLU = clipped_relu.ClippedReLU
```
Comment (unnonouno) [inline chainer/functions/__init__.py]: Please sort alphabetical order

File: chainer/functions/__init__.py [hunk 16]
```diff
@@ -82,6 +84,7 @@
 
 concat = concat.concat
 copy = copy.copy
+clipped_relu = clipped_relu.clipped_relu
```
Comment (unnonouno) [inline chainer/functions/__init__.py]: Please sort alphabetical order

File: chainer/functions/clipped_relu.py [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+from chainer import function
+from chainer.utils import type_check
+import numpy
+
+
+class ClippedReLU(function.Function):
+
+    """Clipped Rectifier Unit function.
+
+    Clipped ReLU is written as :math:`ClippedReLU(x, z) = \min(\max(0, x) ,z)`,
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: Fix position of ",".
`\min(\max(0, x) ,z)` -> `\min(\max(0, x), z)`

File: chainer/functions/clipped_relu.py [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    where :math:`z(>0)` is a parameter to cap return value of ReLU.
+
+    """
+
+    def __init__(self, z):
+        if not isinstance(z, float):
+            raise TypeError('z must be float value')
+        # z must be positive.
+        assert(z > 0)
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: Remove parens

File: chainer/functions/clipped_relu.py [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+            out_types.size() == 1
+        )
+        x_type, = in_types
+        y_type, = out_types
+        type_check.expect(
+            y_type.dtype == numpy.float32,
+            y_type.ndim == x_type.ndim)
+
+    def forward_cpu(self, x):
+        return numpy.array(numpy.minimum(numpy.maximum(0, x[0]), self.cap),
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: Please use `utils.force_array(...)`

File: chainer/functions/clipped_relu.py [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+                           dtype=numpy.float32),
+
+    def backward_cpu(self, x, gy):
+        return numpy.array(gy[0] * (0 < x[0]) * (x[0] < self.cap)),
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: Please use `utils.force_array(...)`

File: chainer/functions/clipped_relu.py [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+            'float* gx, const float* x, const float* gy, const float z',
+            'gx[i] = ((x[i] > 0) and (x[i] < z))? gy[i] : 0',
+            'clipped_relu_bwd')(gx, x[0], gy[0], self.cap)
+        return gx,
+
+
+def clipped_relu(x, z=20.0):
+    """Clipped Rectifier Unit function.
+
+    This function is expressed as :math:`CReLU(x, z) = \min(\max(0, x), z),
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: A backquote is missing

File: chainer/functions/clipped_relu.py [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+    where :math:`z(>0)` is a clipping value.
+
+    Args:
+        x (~chainer.Variable): Input variable, which is n(>0)-dimensional array
+        z (float): clipping value. (default = 20.0)
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: Capitalize "c"

File: chainer/functions/clipped_relu.py [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+
+
+def clipped_relu(x, z=20.0):
+    """Clipped Rectifier Unit function.
+
+    This function is expressed as :math:`CReLU(x, z) = \min(\max(0, x), z),
+    where :math:`z(>0)` is a clipping value.
+
+    Args:
+        x (~chainer.Variable): Input variable, which is n(>0)-dimensional array
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: Remove "(>0)"

**Comment by jnishi** [comment]
Thanks for a lot of comments. I fixes type of docstrings, using utils.force_array, and order of importing.  Please check again.

**Comment by unnonouno** [comment]
OK, finally please fix `docs/functions.rst`.

**Comment by jnishi** [comment]
I added clipped relu to reference manual. Please check again.

File: chainer/functions/clipped_relu.py [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+        # z must be positive.
+        assert z > 0
+        self.cap = z
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 1)
+        x_type, = in_types
+        type_check.expect(x_type.dtype == numpy.float32)
+
+    def check_type_backward(self, in_types, out_types):
```
Comment (unnonouno) [inline chainer/functions/clipped_relu.py]: Sorry this method is deprecated. Please remove it.

**Comment by jnishi** [comment]
I removed check_type_backward. Please check again.

**Comment by unnonouno** [comment]
Great work!! Thank you.

**Pull request merged**